features/79-structured-logging/implementation.md

```diff
@@ -0,0 +1,119 @@
+
+- Log function to print structured logs
+- skygear-server create id per request and pass request id to plugin through context
+- Review all logs and add tag by its functional group (auth, push, pubsub ..etc)
+- Output JSON formatted logs on the skygear cloud. Output readable log when deployed in docker compose. Skygear Cloud can supply custom environment variable to enable this behavior.
+
+#### py-skygear, skygear-SDK-JS
+
+- Log function to print structured logs
+- Inject request id when printing structured logs
+- Review all logs and reduce redundant logs in all plugins, adding corresponding plugin tag
+- Output JSON formatted logs on the skygear cloud. Output readable log when deployed in docker compose. Skygear Cloud can supply custom environment variable to enable this behavior.
+
+### Portal
+
+- Filtering will be done in client side
+- Group logs by request ID option, those logs without request id will not be shown in this mode
+- With log level coloring
+- Increase logs size (Current: 1000)
+
+### skycli
+
+- Filtering will be done in client side, filters are passed by arguments
+- With log level coloring
+
+```sh
+skycli logs
+
+Options:
+  --level       Apply log level filter, comma separated. Support values: debug,info,warning,error,critical
+  --process     Apply process filter, comma separated. Support values: skygear-server,python,js
+  --tag         Apply tag filter, comma separated. Support values: request,auth,db,pubsub,push,auth_plugin,chat_plugin,cms_plugin,sso_plugin,others
+
+Examples:
+  # select skygear server logs and log level is debug
+  skycli --level=debug --process=skygear-server logs
+
+  # select cms and error logs
+  skycli --tags=cms,others logs
+
+  # select cloud functions logs
+  skycli --tags=others logs
+````
+
+### Cloud
+
+- Need to be able to retrieve logs from completed pod
+
+### Structured log format
+
+```js
+{
+    "time": 1521712900.511476,
+    "level": "info", // debug / info / warn / error / critical
+    "process": "server",  // skygear-server / python / js
+    "tag": "db", // request / auth / db / pubsub / push / auth_plugin / chat_plugin / cms_plugin / sso_plugin
+    "request_id": "uuid",
+    "message": "Executed SQL successfully with sql.Queryx",
+    "extra": { // extra information of logs, optional
+        "args": [],
+        "error": "<nil>",
+        "executionCount": 2,
+        "sql": "SELECT record_type, record_field, user_role, writable, readable, comparable, discoverable FROM \"app_chatdemoapp\".\"_record_field_access\"",
+    }
+}
+```
+
+### Python log function
+
+To keep the
+- We can use filter to inject current request id
+https://docs.python.org/3/howto/logging-cookbook.html#using-filters-to-impart-contextual-information
+- Use the logger name for the tag, prefix the name with `tag.`. So if user create their own logger
+in their cloud functions. We can recognized the logs are from cloud functions.
+
+```py
+import logging
+
+logger = logging.getLogger('tag.chat_plugin')
```

Comment:
We have little control over the logger name, for example, sqlalchemy sets its own logger name, so we cannot tell the sqlalchemy log generated from the chat plugin.

My recommendation is that logger name can be anything, and we do not apply semantic to what logger name means. The logger name is included in the log output as `logger="[loggername]"` field.

The `tag` is added to the log record using log filtering (see the python doc that carmen mentioned 👍 ) and supply the `tag` using thread-local variable. The thread-local variable is set at the time when we receive the request.

For example:

1. Server receive call to `chat:send_message` (the name of the lambda is for illustration purpose (i.e. not real name))
2. Python calls `chat:send_message` lambda.
3. In the lambda, we set `tag=chat` as thread local variable.
4. Lambda call log function
5. Filter get `tag=chat` from thread local variable and add it to log record.

For user lambda, we can have `tag=user` by default, so that the user doesn’t need to know about tag when the user use log functions. skycli/portal can filter based on `tag=user` tag.

I hope this makes the implementation easier as we don’t need to interpret the logger name, at the same time we support the use case of separating user log from plugin logs.